cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -493,6 +500,16 @@ else()
     )
 endif()
 
+if(DEFINED ENV{UTF8PROC_SOURCE_URL})
```

Comment:
Could you use `ARROW_UTF8PROC_URL` like other packages?
(Could you also fix `BZIP2_SOURCE_URL`?)